cellengine/resources/compensation.py

```diff
@@ -1,68 +1,95 @@
-from dataclasses import dataclass, field
 from typing import List, TYPE_CHECKING
+from attr import define, field
 
-from dataclasses_json.cfg import config
-import numpy
+from numpy import array, linalg
 from pandas import DataFrame
 
 import cellengine as ce
-from cellengine.utils.dataclass_mixin import DataClassMixin, ReadOnly
+from cellengine.utils import readonly
+from cellengine.utils.converter import converter
 
 
 if TYPE_CHECKING:
     from cellengine.resources.fcs_file import FcsFile
 
 
-@dataclass
-class Compensation(DataClassMixin):
+@define
+class Compensation:
     """A class representing a CellEngine compensation matrix.
 
     Can be applied to FCS files to compensate them.
     """
 
+    _id: str = field(on_setattr=readonly)
+    experiment_id: str = field(on_setattr=readonly)
     name: str
     channels: List[str]
-    dataframe: DataFrame = field(
-        metadata=config(
-            field_name="spillMatrix",
-            encoder=lambda x: x.to_numpy().flatten().tolist(),
-            decoder=lambda x: numpy.array(x),
-        )
-    )
-    _id: str = field(
-        metadata=config(field_name="_id"), default=ReadOnly()
-    )  # type: ignore
-    experiment_id: str = field(default=ReadOnly())  # type: ignore
-
-    def __post_init__(self):
-        self.dataframe = DataFrame(
-            self.dataframe.reshape(self.N, self.N),
+    spill_matrix: List[float]
+
+    @property
+    def dataframe(self):
+        return DataFrame(
+            array(self.spill_matrix).reshape(self.N, self.N),  # type: ignore
             columns=self.channels,
             index=self.channels,
         )
 
+    @dataframe.setter
+    def dataframe(self, val: DataFrame):
+        try:
+            assert len(val.columns) == len(val.index)
+            assert all(val.columns == val.index)
+            self.channels = val.columns.to_list()
+            self.spill_matrix = val.to_numpy().flatten().tolist()
+        except Exception as e:
+            raise e
+
     def __repr__(self):
         return f"Compensation(_id='{self._id}', name='{self.name}')"
 
     @property
-    def N(self):
-        return len(self.channels)
+    def path(self):
+        return f"experiments/{self.experiment_id}/compensations/{self._id}".rstrip(
+            "/None"
+        )
+
+    @classmethod
+    def from_dict(cls, data: dict):
+        return converter.structure(data, cls)
+
+    def to_dict(self):
+        return converter.unstructure(self)
 
     @classmethod
     def get(cls, experiment_id: str, _id: str = None, name: str = None) -> Compensation:
         kwargs = {"name": name} if name else {"_id": _id}
         return ce.APIClient().get_compensation(experiment_id, **kwargs)
 
     @classmethod
-    def create(cls, experiment_id: str, compensation: dict) -> Compensation:
-        """Creates a compensation
+    def create(
+        cls,
+        experiment_id: str,
+        name: str,
+        channels: List[str],
+        spill_matrix: List[float],
```

Comment:
Later: it would be nice to accept a DataFrame in lieu of `channels` and the flat `spill_matrix`.

Reply:
How do you mean? Currently a user can update `comp.dataframe` and that will update the channels and spill_matrix of the comp

Reply:
Right now you'd have to do:
```py
comp = Compensation.create(experiment_id, "my comp", [], [])
comp.dataframe = df
```
whereas I'm suggesting allowing
```py
comp = Compensation.create(experiment_id, "my comp", df)
```

Reply:
aaaaaah okay, gotcha

Reply:
added :sunny: